Account/FxADeviceRegistrator.swift

```diff
@@ -0,0 +1,125 @@
+        }
+
+        let client = FxAClient10(endpoint: account.configuration.authEndpointURL)
+        let sessionToken = state.sessionToken
+        let name = DeviceInfo.defaultClientName()
+        let device: FxADevice
+        if let deviceId = account.fxaDeviceId {
+            device = FxADevice.forUpdate(name, id: deviceId)
+        } else {
+            device = FxADevice.forRegister(name, type: "mobile")
+        }
+
+        return client.registerOrUpdateDevice(sessionToken, device: device).bind { result in
+            let deferred = Deferred<Maybe<FxADeviceRegistrationResult>>()
+
+            if let device = result.successValue,
+                let deviceId = device.id {
+                account.fxaDeviceId = deviceId
+                account.deviceRegistrationVersion = DeviceRegistrationVersion
+                deferred.fill(Maybe(success: FxADeviceRegistrationResult.RegisteredOrUpdated))
+                return deferred
+            }
+
+            let error = result.failureValue as? FxAClientError
+            switch (error) {
+            case let .Remote(remoteError)?:
+                switch (remoteError.code) {
+                case FxAccountRemoteError.DEVICE_SESSION_CONFLICT:
+                    recoverFromDeviceSessionConflict(account, client: client, deferred: deferred, sessionToken: sessionToken)
+                case FxAccountRemoteError.UNKNOWN_DEVICE:
+                    // TODO: Shouldn't this also clear registration version? Android doesn't.
+                    handleUnknownDevice(account)
+                case FxAccountRemoteError.INVALID_AUTHENTICATION_TOKEN:
+                    logErrorAndResetDeviceRegistrationVersion(account, description: String(remoteError))
+                    handleTokenError(account, client: client)
+                default:
+                    logErrorAndResetDeviceRegistrationVersion(account, description: String(remoteError))
+                }
+            case let .Local(localError)?:
+                self.logErrorAndResetDeviceRegistrationVersion(account, description: localError.description)
+            default:
+                // This shouldn't happen.
+                self.logErrorAndResetDeviceRegistrationVersion(account, description: "invalid error type")
+                assertionFailure()
+            }
+
+            deferred.fillIfUnfilled(Maybe(failure: DeviceRegistratorError.DeviceRegistrationFailed))
+
+            return deferred
+        }
+    }
+
+    private static func handleUnknownDevice(account: FirefoxAccount) {
+        log.info("unknown device id, clearing the cached device id")
+        account.fxaDeviceId = nil
+    }
+
+    private static func recoverFromDeviceSessionConflict(account: FirefoxAccount, client: FxAClient10, deferred: Deferred<Maybe<FxADeviceRegistrationResult>>, sessionToken: NSData) {
+        log.warning("device session conflict, attempting to ascertain the correct device id")
+        client.devices(sessionToken).upon { response in
```

Comment:
The `client.devices` method is async but when this is called in the above error handling code it makes it look synchronous. I can see this causing race conditions because we're returning from the error handling potentially before this code finishes. We should make sure to return a deferred here to chain after to keep the timing straight.